### PR TITLE
Gulp to Webpack + offline functionality

### PR DESCRIPTION
**Pull request by paulxuca**
it's 207 season boys

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -14,6 +14,7 @@ To install the required dependencies, simply run the following commands in the r
 
 ```shell
 $ npm install
+$ yarn install
```
Comment (arkon) [inline README.md]: I'd rather leave this out.

File: package.json [hunk 2]
```diff
@@ -10,30 +10,31 @@
...
-    "del": "^2.2.2",
-    "gulp": "github:gulpjs/gulp#4.0",
-    "gulp-cssnano": "^2.1.2",
-    "gulp-gh-pages": "^0.5.4",
-    "gulp-htmlmin": "^3.0.0",
-    "gulp-sass": "^2.3.2",
-    "gulp-uglify": "^2.0.0",
-    "vinyl-buffer": "^1.0.0",
-    "vinyl-source-stream": "^1.1.0"
+    "babel-preset-stage-0": "^6.16.0",
```
Comment (arkon) [inline package.json]: Why? We only really need ES2015 + the transform-class-properties plugin.

File: src/script.js [hunk 3]
```diff
@@ -1,6 +1,31 @@
...
+  <AppContainer>
+    <CRCMaker />
+  </AppContainer>,
+document.getElementById('main'));
+
+// REACT HMR
+if (module.hot) {
+  module.hot.accept('./components/CRCMaker', () => {
+    const NextHotLoaded = require('./components/CRCMaker.js').default; // eslint-disable-line
+    ReactDOM.render(
```
Comment (arkon) [inline src/script.js]: Just use `render`.

File: .gitignore [hunk 4]
```diff
@@ -1,5 +1,7 @@
 .publish/
-build/
 etc/
 node_modules/
 npm-debug.log
+dist/
+cert.pem
```
Comment (arkon) [inline .gitignore]: Maybe include instructions on how to generate these?

File: webpack/webpack.prod.config.js [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+      loader: 'babel-loader',
+      query: {
+        presets: ['es2015', 'stage-0', 'react'],
+      },
+    }, {
+      test: /\.scss$/,
+      exclude: /node_modules/,
+      loader: ExtractTextPlugin.extract({
+        fallbackLoader: 'style-loader',
+        loader: 'css!sass?indentedSyntax=true&sourceMap=true',
```
Comment (arkon) [inline webpack/webpack.prod.config.js]: No need for indented syntax or sourcemaps in prod.

**Comment by paulxuca** [comment]
added changes in! there's some code for service workers for #32, which isn't done yet (having trouble testing it without a https connection)

File: webpack/webpack.prod.config.js [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+    filename: 'app.[hash].js',
+  },
+  devtool: 'eval',
+  module: {
+    loaders:[{
+      test: /\.js$/,
+      include: path.resolve(process.cwd(), 'src'),
+      loader: 'babel-loader',
+      query: {
+        presets: ['es2015', 'react'],
```
Comment (arkon) [inline webpack/webpack.prod.config.js]: Don't we need the plugin for transpiling the static class variables?

File: package.json [hunk 7]
```diff
@@ -10,30 +10,30 @@
   "license": "MIT",
   "private": true,
   "scripts": {
-    "build": "gulp",
-    "watch": "gulp watch"
+    "dev": "node ./webpack.server.js",
+    "build": "rimraf ./dist && webpack --config ./webpack/webpack.prod.config.js",
+    "preview": "http-server ./dist -c-1 -o"
```
Comment (arkon) [inline package.json]: So does this work? You could try using [simplehttp2server](https://github.com/GoogleChrome/simplehttp2server), but you still need to generate some certificates locally.

File: src/script.js [hunk 8]
```diff
@@ -1,6 +1,31 @@
...
 import CRCMaker from './components/CRCMaker';
 
-render(<CRCMaker />, document.getElementById('main'));
+require('./styles/styles.scss');
+
+render(
+  <AppContainer>
+    <CRCMaker />
+  </AppContainer>,
+document.getElementById('main'));
```
Comment (arkon) [inline src/script.js]: Indent

File: src/script.js [hunk 9]
```diff
@@ -1,6 +1,31 @@
...
+
+// REACT HMR
+if (module.hot) {
+  module.hot.accept('./components/CRCMaker', () => {
+    const NextHotLoaded = require('./components/CRCMaker.js').default; // eslint-disable-line
+    render(
+      <AppContainer>
+        <NextHotLoaded />
+      </AppContainer>
+      ,
```
Comment (arkon) [inline src/script.js]: Move comma up

File: webpack.server.js [hunk 10]
```diff
@@ -0,0 +1,16 @@
+var webpack = require('webpack');
```
Comment (arkon) [inline webpack.server.js]: Use `const`

File: webpack/webpack.dev.config.js [hunk 11]
```diff
@@ -0,0 +1,38 @@
+var path = require('path');
```
Comment (arkon) [inline webpack/webpack.dev.config.js]: Use `const`

File: webpack/webpack.prod.config.js [hunk 12]
```diff
@@ -0,0 +1,69 @@
+var path = require('path');
```
Comment (arkon) [inline webpack/webpack.prod.config.js]: Use `const`

**Comment by paulxuca** [comment]
should be good now! 👍

**Comment by arkon** [comment]
@paulxuca Looks like there's a merge conflict since master's been updated in the meantime.

File: package.json [hunk 13]
```diff
@@ -10,30 +10,32 @@
...
+    "html-webpack-plugin": "^2.24.0",
+    "http-server": "^0.9.0",
+    "node-sass": "^3.10.1",
+    "offline-plugin": "^3.4.2",
+    "react-hot-loader": "^3.0.0-beta.6",
+    "rimraf": "^2.5.4",
+    "sass-loader": "^4.0.2",
+    "webpack": "2.1.0-beta.22",
+    "webpack-dev-server": "2.1.0-beta.9",
+    "webpack-merge": "^0.15.0",
```
Comment (arkon) [inline package.json]: Looks like you rolled your own way of merging webpack configs, so I guess this isn't required.

File: package.json [hunk 14]
```diff
@@ -10,30 +10,32 @@
...
+    "webpack-validator": "^2.2.9"
```
Comment (arkon) [inline package.json]: I don't think this is used anywhere.

File: package.json [hunk 15]
```diff
@@ -10,30 +10,32 @@
...
-    "vinyl-source-stream": "^1.1.0"
+    "extract-text-webpack-plugin": "v2.0.0-beta.4",
+    "html-webpack-plugin": "^2.24.0",
+    "http-server": "^0.9.0",
+    "node-sass": "^3.10.1",
+    "offline-plugin": "^3.4.2",
+    "react-hot-loader": "^3.0.0-beta.6",
+    "rimraf": "^2.5.4",
+    "sass-loader": "^4.0.2",
+    "webpack": "2.1.0-beta.22",
```
Comment (arkon) [inline package.json]: Latest version of webpack is 2.1.0-beta.25.

File: package.json [hunk 16]
```diff
@@ -10,30 +10,32 @@
...
-    "del": "^2.2.2",
-    "gulp": "github:gulpjs/gulp#4.0",
-    "gulp-cssnano": "^2.1.2",
-    "gulp-gh-pages": "^0.5.4",
-    "gulp-htmlmin": "^3.0.0",
-    "gulp-sass": "^2.3.2",
-    "gulp-uglify": "^2.0.0",
-    "vinyl-buffer": "^1.0.0",
-    "vinyl-source-stream": "^1.1.0"
+    "extract-text-webpack-plugin": "v2.0.0-beta.4",
```
Comment (arkon) [inline package.json]: Get rid of the `v`

File: src/script.js [hunk 17]
```diff
@@ -1,6 +1,30 @@
...
+require('./styles/styles.scss');
+
+render(
+  <AppContainer>
+    <CRCMaker />
+  </AppContainer>,
+  document.getElementById('main'));
+
+// REACT HMR
+if (module.hot) {
```
Comment (arkon) [inline src/script.js]: Maybe also put `process.env.NODE_ENV === 'development'` or something so it'll be completely removed when built for prod?

**Comment by paulxuca** [comment]
there we go!

File: src/index.html [hunk 18]
```diff
@@ -55,6 +54,4 @@ <h1 class="header__title">CRC Card Maker</h1>
       location.reload();
     };
   </script>
-  <script async src="script.js"></script>
-</body>
```
Comment (arkon) [inline src/index.html]: Wait, we still need the closing `</body>` tag!

**Comment by paulxuca** [comment]
hopefully that's it 📦

**Comment by arkon** [comment]
Thanks @paulxuca !

**Comment by arkon** [comment]
Okay apparently Webpack doesn't play friendly with the PDF stuff at all, so I've moved it to a separate branch for now. I'll try to fix it later.

**Comment by arkon** [comment]
Finally merged this stuff into master properly. 👍

**Pull request merged**